### PR TITLE
Add feature to help people move to using primary constructors.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Given code like so:

image

It offers:

image

The first produces code like so (keeping the members around):

image

The second produces code like so:

image

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
todo: 

- [ ] disable if fields have attributes on them.
- [ ] add test that shows we don't remove members if a value would bind to a different member.

File: src/Analyzers/CSharp/Analyzers/UsePrimaryConstructor/CSharpUsePrimaryConstructorDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticDocumentAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(context =>
+            {
+                // "x is not Type y" is only available in C# 9.0 and above. Don't offer this refactoring
+                // in projects targeting a lesser version.
+                if (!context.Compilation.LanguageVersion().IsCSharp12OrAbove())
```
Comment (DoctorKrolic) [inline src/Analyzers/CSharp/Analyzers/UsePrimaryConstructor/CSharpUsePrimaryConstructorDiagnosticAnalyzer.cs:77]: Fix comment here)

File: src/Analyzers/Core/Analyzers/IDEDiagnosticIds.cs [hunk 2]
```diff
@@ -190,6 +190,8 @@ internal static class IDEDiagnosticIds
 
         public const string UseNameofInAttributeDiagnosticId = "IDE0280";
 
+        public const string UsePrimaryConstructorDiagnosticId = "IDE0290";
```
Comment (DoctorKrolic) [inline src/Analyzers/Core/Analyzers/IDEDiagnosticIds.cs:193]: Just curious: why modern `IDE` diagnostic ids are spread out by 10?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/LanguageVersionExtensions.cs [hunk 3]
```diff
@@ -12,6 +12,9 @@ public static bool IsCSharp11OrAbove(this LanguageVersion languageVersion)
         public static bool HasConstantInterpolatedStrings(this LanguageVersion languageVersion)
             => languageVersion >= LanguageVersion.CSharp10;
 
+        public static bool IsCSharp12OrAbove(this LanguageVersion languageVersion)
+            => languageVersion >= LanguageVersion.Preview;
```
Comment (DoctorKrolic) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/LanguageVersionExtensions.cs:16]: Maybe some `TODO` here for replacing this with actual lang version 12 when it is actually arrives?

File: src/Analyzers/Core/Analyzers/IDEDiagnosticIds.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Analyzers/Core/Analyzers/IDEDiagnosticIds.cs:193]: It's reasonably common to add more diagnostics to a feature in the future. This allows the ide to keep those later diagnostics associated together. :-)

File: src/Analyzers/CSharp/CodeFixes/UsePrimaryConstructor/CSharpUsePrimaryConstructorCodeFixProvider.cs [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+            context.RegisterCodeFix(
+                CodeAction.Create(
+                    CSharpAnalyzersResources.Use_primary_constructor,
+                    cancellationToken => UsePrimaryConstructorAsync(document, constructorDeclaration, properties, removeMembers: false, cancellationToken),
+                    nameof(CSharpAnalyzersResources.Use_primary_constructor)),
+                diagnostic);
+
+            if (diagnostic.Properties.Count > 0)
+            {
+                var (resource, equivalenceKey) =
```
Comment (akhera99) [inline src/Analyzers/CSharp/CodeFixes/UsePrimaryConstructor/CSharpUsePrimaryConstructorCodeFixProvider.cs:73]: Is the gif outdated?

Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/CodeFixes/UsePrimaryConstructor/CSharpUsePrimaryConstructorCodeFixProvider.cs:73]: yes.

**Review by akhera99 — APPROVED** [review]

File: src/Analyzers/CSharp/CodeFixes/UsePrimaryConstructor/CSharpUsePrimaryConstructorCodeFixProvider.cs [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+
+    public override ImmutableArray<string> FixableDiagnosticIds
+        => ImmutableArray.Create(IDEDiagnosticIds.UsePrimaryConstructorDiagnosticId);
+
+    public override FixAllProvider? GetFixAllProvider()
+#if CODE_STYLE
+        => WellKnownFixAllProviders.BatchFixer;
+#else
+        => new CSharpUsePrimaryConstructorFixAllProvider();
+
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UsePrimaryConstructor/CSharpUsePrimaryConstructorCodeFixProvider.cs:48]:
```suggestion
```

**Pull request merged**